std/algebra/native/sw_bls12377/g1.go

```diff
@@ -671,3 +672,208 @@ func (P *G1Affine) scalarBitsMul(api frontend.API, Q G1Affine, s1bits, s2bits []
 
 	return P
 }
+
+// fake-GLV
+func (R *G1Affine) scalarMulGLVAndFakeGLV(api frontend.API, P G1Affine, s frontend.Variable, opts ...algopts.AlgebraOption) *G1Affine {
+	cfg, err := algopts.NewConfig(opts...)
+	if err != nil {
+		panic(err)
+	}
+	cc := getInnerCurveConfig(api.Compiler().Field())
+
+	// handle zero-scalar
+	var selector0 frontend.Variable
+	_s := s
+	if cfg.CompleteArithmetic {
+		selector0 = api.IsZero(s)
+		_s = api.Select(selector0, 1, s)
+	}
+
+	// Instead of computing [s]P=Q, we check that Q-[s]P == 0.
+	// Checking Q - [s]P = 0 is equivalent to [v]Q + [-s*v]P = 0 for some nonzero v.
+	//
+	// The GLV curves supported in gnark have j-invariant 0, which means the eigenvalue
+	// of the GLV endomorphism is a primitive cube root of unity.  If we write
+	// v, s and r as Eisenstein integers we can express the check as:
+	//
+	// 			[v1 + λ*v2]Q + [u1 + λ*u2]P = 0
+	// 			[v1]Q + [v2]phi(Q) + [u1]P + [u2]phi(P) = 0
+	//
+	// where (v1 + λ*v2)*(s1 + λ*s2) = u1 + λu2 mod (r1 + λ*r2)
+	// and u1, u2, v1, v2 < r^{1/4} (up to a constant factor).
+	//
+	// This can be done as follows:
+	// 		1. decompose s into s1 + λ*s2 mod r s.t. s1, s2 < sqrt(r) (hinted classical GLV decomposition).
+	// 		2. decompose r into r1 + λ*r2  s.t. r1, r2 < sqrt(r) (hardcoded half-GCD of λ mod r).
+	// 		3. find u1, u2, v1, v2 < c*r^{1/4} s.t. (v1 + λ*v2)*(s1 + λ*s2) = (u1 + λ*u2) mod (r1 + λ*r2).
+	// 		   This can be done through a hinted half-GCD in the number field
+	// 		   K=Q[w]/f(w).  This corresponds to K being the Eisenstein ring of
+	// 		   integers i.e. w is a primitive cube root of unity, f(w)=w^2+w+1=0.
+	//
+	// The hint returns u1, u2, v1, v2 and the quotient q.
+	// In-circuit we check that (v1 + λ*v2)*s = (u1 + λ*u2) + r*q
+	sd, err := api.NewHint(halfGCDEisenstein, 5, _s, cc.lambda)
+	if err != nil {
+		panic(fmt.Sprintf("halfGCDEisenstein hint: %v", err))
+	}
+	u1, u2, v1, v2, q := sd[0], sd[1], sd[2], sd[3], sd[4]
+
+	// Eisenstein integers real and imaginary parts can be negative. So we
+	// return the absolute value in the hint and negate the corresponding
+	// points here when needed.
+	signs, err := api.NewHint(halfGCDEisensteinSigns, 5, _s, cc.lambda)
+	if err != nil {
+		panic(fmt.Sprintf("halfGCDEisensteinSigns hint: %v", err))
+	}
+	isNegu1, isNegu2, isNegv1, isNegv2, isNegq := signs[0], signs[1], signs[2], signs[3], signs[4]
+
+	// We need to check that:
+	// 		s*(v1 + λ*v2) + u1 + λ*u2 - r * q = 0
```

Comment:
Here we would need to keep `q` (referencing to the comment in `sw_emulated`, where I recommended removing it).

But now don't we have the problem of the overflow? Here `s` can be 253 bits, and all `v1, v2, u1, u2, lambda` 64(? or more with constants) bits. Now, when computing `s*(v1 + lambda*v)` the result may be ~443 (my rought estimate) bits, which overflows the BW6 scalar field.

I'd think that decomposing `s` actually works better here, imo we don't overflow then, but I'm not sure.

Reply:
makes sense. Actually in this PR we are not using `scalarMulGLVAndFakeGLV` anywhere except for cross testing other methods. It's currently (with +10 bits) more expensive than classical GLV and doing the scalars decomposition with emulated arithmetic would make it even more expensive. So not using it for native package.